pkg/util/volume_test.go

```diff
@@ -0,0 +1,219 @@
+						Name: name,
+					},
+				},
+				&api.Volume{
+					Id: "two",
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+				},
+			}, nil).
+			Times(1),
+		driver.
+			EXPECT().
+			Enumerate(&api.VolumeLocator{Name: name}, nil).
+			Return([]*api.Volume{}, nil).
+			Times(1),
+	)
+
+	// Expect not found
+	_, err := VolumeFromName(driver, name)
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Cannot locate")
+
+	// Setup calls
+	gomock.InOrder(
+		// Return that it was not found
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+
+		// Return too many volumes
+		driver.
+			EXPECT().
+			Enumerate(&api.VolumeLocator{Name: name}, nil).
+			Return([]*api.Volume{
+				&api.Volume{
+					Id: "one",
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+				},
+				&api.Volume{
+					Id: "two",
+					Locator: &api.VolumeLocator{
+						Name: name,
+					},
+				},
+			}, nil).
+			Times(1),
+	)
+
+	// Expect not found
+	_, err = VolumeFromName(driver, name)
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Cannot locate")
+
+	// Setup calls
+	gomock.InOrder(
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+
+		driver.
+			EXPECT().
+			Enumerate(&api.VolumeLocator{Name: name}, nil).
+			Return(nil, fmt.Errorf("Error")).
+			Times(1),
+	)
+
+	// Expect not found
+	_, err = VolumeFromName(driver, name)
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Failed to locate")
+}
+
+func TestVolumeFromNameFailedToLocate(t *testing.T) {
+
+	mc := gomock.NewController(t)
+	defer mc.Finish()
+
+	driver := mock.NewMockVolumeDriver(mc)
+
+	// Setup calls
+	name := "myvolume"
+	gomock.InOrder(
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+		driver.
+			EXPECT().
+			Enumerate(&api.VolumeLocator{Name: name}, nil).
+			Return(nil, fmt.Errorf("not found")).
+			Times(1),
+	)
+
+	// Expect not found
+	_, err := VolumeFromName(driver, name)
+	assert.NotNil(t, err)
+}
+
+func TestVolumeFromNameFoundFromInspect(t *testing.T) {
+
+	mc := gomock.NewController(t)
+	defer mc.Finish()
+
+	driver := mock.NewMockVolumeDriver(mc)
+
+	// Setup calls
+	name := "myvolume"
+	driver.
+		EXPECT().
+		Inspect([]string{name}).
+		Return([]*api.Volume{
+			&api.Volume{
+				Id: name,
+				Locator: &api.VolumeLocator{
+					Name: "hello",
+				},
+			},
+		}, nil).
+		Times(1)
+
+	// Expect not found
+	v, err := VolumeFromName(driver, name)
+	assert.Nil(t, err)
+	assert.NotNil(t, v)
+	assert.Equal(t, v.Id, name)
+	assert.Equal(t, v.GetLocator().GetName(), "hello")
+}
+
+func TestVolumeFromNameFoundFromEnumerate(t *testing.T) {
+
+	mc := gomock.NewController(t)
+	defer mc.Finish()
+
+	driver := mock.NewMockVolumeDriver(mc)
+
+	// Setup calls
+	name := "myvolume"
+	gomock.InOrder(
+		driver.
+			EXPECT().
+			Inspect([]string{name}).
```

Comment:
Add test for Inspect(nil) and Inspect([]string{}) if not already.

Reply:
Wouldn't that mean that someone has to pass nil into Inspect()? Could you show me an example of the code in CSI where that would happen?

Reply:
If it can be done, it needs to be handled, regardless of someone is doing it today. Can you guarantee that no one in future will do it ?

Reply:
I think I am confused because I am not testing Inspect(), or expecting to have inspect be passed with nil. If the unit test passes nil to Inspect() then it will fail, since it expected something else. I think what you want to be tested is already being tested, right?